modules/core/src/lib/layer.js

```diff
@@ -694,6 +702,14 @@ ${flags.viewportChanged ? 'viewport' : ''}\
       attributeManager && attributeManager.getNeedsRedraw({clearRedrawFlags});
     redraw = redraw || attributeManagerNeedsRedraw;
 
+    redraw = redraw || this._modelNeedsRedraw(clearRedrawFlags);
```

Comment:
Now that the check goes several layers deep, it is no longer trivial. This function should return as soon as `redraw` is not empty.

Reply:
> This function should return as soon as redraw is not empty.

I am open to changing this. However it is a tradeoff between clearing all dirty flags on redraw, or keep redrawing one frame for each dirty object in the scene until all flags are cleared.